### PR TITLE
Add reservoir sampling

### PR DESCRIPTION
**Pull request by shepmaster**
Howdy! This PR is certainly not production ready, but it works and I wanted to get some guidance about how to continue. Some points I have questions about:

1. How should this be tested? Testing random stuff is not my specialty, but I've always done basic statistical analysis. In this case, I did "*do each of the items occur about the same amount of times, with some fuzziness*".
2. How do you feel about including a production dependency? Should it have a Cargo feature flag including it?
3. Does the code fit the libraries coding style?

Obviously, documentation will need to be added.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1260,6 +1263,23 @@ pub trait Itertools : Iterator {
...
+        where Self: Sized
+    {
+        let rng = &mut rand::thread_rng();
+
+        let mut reservoir: Vec<_> = self.by_ref().take(sample_size).collect();
+        for (i, item) in self.enumerate() {
+            let range = Range::new(0, i + sample_size + 1);
+            let replacement_index = range.ind_sample(rng);
+
+            if let Some(spot) = reservoir.get_mut(replacement_index) {
```
Comment (bluss) [inline src/lib.rs]: ah I love this, when the bounds check has a purpose for the algorithm.

**Comment by bluss** [comment]
Hey, reservoir sampling is a cool algorithm, but I think the dependency and so on signals that this is not for itertools. It may be better in another kind of crate. We should try to keep focus on general iterator adaptors, I guess the boundary is fuzzy. We may already have too many special case things included.

As for testing, I'd do it just like you I think. Don't see any other nice way to do it immediately now.

**Comment by shepmaster** [comment]
> another kind of crate

That makes sense. Do you know of any "iterator terminator" crates? In my mind sampling is very similat to functions like `fold` or `min`.

**Comment by bluss** [comment]
oh I meant general iterator adaptors and methods; there are iterator methods / terminators in itertools.

**Comment by shepmaster** [comment]
Closing in favor of rust-lang/rand#72.

Thanks!

**Pull request closed without merging**